sdk/core/azure-core/src/http/retry_policy.cpp

```diff
@@ -135,6 +135,9 @@ std::unique_ptr<RawResponse> Azure::Core::Http::RetryPolicy::Send(
   for (RetryNumber attempt = 1;; ++attempt)
   {
     Delay retryAfter{};
+    request.StartTry();
+    // creates a copy of original query parameters from request
+    auto originalQueryParameters = request.GetUrl().GetQueryParameters();
```

Comment:
We should save the URL string instead using `GetAbsoluteUrl` and restore it later by using the URL ctor.

Reply:
URL ctor parses the string to get each URL part.  We would never change host, port or other thing but the query parameters.
That's why I preferred 2 copy operations than copy + parsing the copy back

Reply:
I understand, but we are letting that implementation choice impact the API design. Can you think of other usages/scenarios where `GetQueryParameters()` would be useful?

We can try to optimize the Url parsing further to minimize that cost, and that way we don't leak an implementation detail of the `Url` class using  map under the covers in public surface area.

Reply:
I agree with @ahsonkhan that we should save the URL.

But `GetQueryParameters` is also useful, when storage sdk signs the request.

Reply:
> But `GetQueryParameters` is also useful, when storage sdk signs the request.

How so, @Jinming-Hu - can you point me to a code snippet on its usage or show a small example using pseudocode?

Reply:
@ahsonkhan , I think is it super useful if I can call this function and iterate each query parameter without any parsing (since the parsing was already done by the URL).

This is how it is used in the shared_key_policy
```cpp
// canonicalized resource
    string_to_sign += "/" + m_credential->AccountName + "/" + request.GetUrl().GetPath() + "\n";
    for (const auto& query : request.GetUrl().GetQueryParameters())
    {
      std::string key = query.first;
      std::transform(key.begin(), key.end(), key.begin(), [](char c) {
        return static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
      });
      ordered_kv.emplace_back(std::make_pair(
          Azure::Core::Http::Url::Decode(key), Azure::Core::Http::Url::Decode(query.second)));
    }
```

If we removed this method, I would need to take the full URL, or at least the relative URL and then parse it myself to read each query parameter.

I don't think we need to explicitly hide the query parameters map from users. I see benefits on making it accessible. Yes, it is true that it breaks encapsulation, but it is worth it 👍  :)

Reply:
> As a follow up is `SetQueryParameters` useful?

`SetQueryParameters` is useful when our customers generate a SAS token on Azure Portal. SAS token is something like 

```
?sv=2019-12-12&ss=bfqt&srt=sco&sp=rwdlacupx&se=2020-09-23T09:48:17Z&st=2020-09-23T01:48:17Z&spr=https&sig=XXXXXXXXX
```

Reply:
Don't they, then, need to create a map from the SAS token to pass into `SetQueryParameters`? How would they do that? Wouldn't they pass that directly to the `Url` ctor instead?

Reply:
We can let the users to parse the sas token (query parameters) themselves, but I don't think it's a good idea

Actually we can let the users do all these stuff (url parsing, encoding, decoding) themselves. Then this URL class becomes useless.

Reply:
> We can let the users to parse the sas token (query parameters) themselves, but I don't think it's a good idea

That's not what I am asking nor suggesting.

> Actually we can let the users do all these stuff (url parsing, encoding, decoding) themselves.

To clarify, I am only asking about a canonical use case of `SetQueryParameters` specifically to understand its need, outside of using it to restore a previous map-based snapshot of the Url query parameters. You mentioned an example, let's dig into that a little bit.

What form does the caller have the SAS token in? Presumably, its a string. Couldn't the caller pass that in to the `Url` ctor and let it do the parsing. I don't understand, yet, why they would need `SetQueryParameters` for this case.

Can you please expand on that and help me understand how we turn the SAS token string into a query parameter map, using this Url class, which is then passed to `SetQueryParameters`? Somehow the user needs to get a map, right? How do they get it?

Reply:
currently we have two places using this `SetQueryParameters`.

1. When we parse connection string, we first get a URL and then get a SAS token. Of course we can avoid using `SetQueryParameters` in this case by manually concatenating SAS token after URL.

2. We have a URL of a resource. This url may already contain a query parameter like `snapshot=` or `versionid=`. We want to append the SAS token to this URL. We cannot replace `SetQueryParameters` with an easy way in this case.

Reply:
@Jinming-Hu , I believe @ahsonkhan  is referring to the overloading method I added to set query parameters directly from a map<string, string>.
Before this new method, `setQueryParameters`  would be possible by receiving one `std::string` that contains all query parameters.

I added this overload method so we can restore the query parameters Map without parsing the entire URL string again.
I can also remove this function retryPolicy a Url's friend class, so we can set the private Map directly and this keeps hidden from customers...
However, we expect Services to create its own version of a Retry Policy (like the one from Storage), and they might also find useful to have the access the query parameters Map.

Even after removing the access, we would not hide the Map from customers, since we are offering `getQueryParameters` function and we want to return the Map in there (actually now we are returning a ref to the map).

Reply:
Well, I don't think the overloading `SetQueryParameters` that accepts `std::map` is necessary.

Reply:
and also like I already said in this thread, I think we should save the whole URL rather than just query parameters on retry.

Reply:
> is referring to the overloading method I added to set query parameters directly from a map<string, string>.

Exactly. Thanks for making that clear, @vhvb1989 

I am referring to these, and we should consider removing them, at least for now. That said, @vhvb1989, I will leave this for you to decide. 
```cpp
void SetQueryParameters(std::map<std::string, std::string> queryParameters)
const std::map<std::string, std::string>& GetQueryParameters() const
```

One thing you can consider, is measure what the duplicated "parse" logic costs when using strings as snapshots compared to the cost of the map iteration and decoding (that you showed in the shared_key_policy example). I suspect, the parsing logic won't be too significant.

> Well, I don't think the overloading `SetQueryParameters` that accepts `std::map` is necessary.
> I think we should save the whole URL rather than just query parameters on retry.

+1